homeassistant/components/xiaomi_miio/config_flow.py

```diff
@@ -47,35 +51,73 @@ async def async_step_user(self, user_input=None):
+
+            unique_id = f"{model}-{mac_address}-gateway"
+            await self.async_set_unique_id(unique_id)
+            self._abort_if_unique_id_configured()
+
+            return await self.async_step_gateway()
+
+        # Discovered device is not yet supported
+        return self.async_abort(reason="not_xiaomi_miio")
+
     async def async_step_gateway(self, user_input=None):
         """Handle a flow initialized by the user to configure a gateway."""
         errors = {}
         if user_input is not None:
-            host = user_input[CONF_HOST]
             token = user_input[CONF_TOKEN]
-
-            # Try to connect to a Xiaomi Gateway.
-            connect_gateway_class = ConnectXiaomiGateway(self.hass)
-            await connect_gateway_class.async_connect_gateway(host, token)
-            gateway_info = connect_gateway_class.gateway_info
-
-            if gateway_info is not None:
-                unique_id = f"{gateway_info.model}-{gateway_info.mac_address}-gateway"
-                await self.async_set_unique_id(unique_id)
-                self._abort_if_unique_id_configured()
-                return self.async_create_entry(
-                    title=user_input[CONF_NAME],
-                    data={
-                        CONF_FLOW_TYPE: CONF_GATEWAY,
-                        CONF_HOST: host,
-                        CONF_TOKEN: token,
-                        "gateway_id": unique_id,
-                        "model": gateway_info.model,
-                        "mac": gateway_info.mac_address,
-                    },
-                )
-
-            errors["base"] = "connect_error"
+            if user_input.get(CONF_HOST):
+                self.host = user_input[CONF_HOST]
+
+            if self.host:
+                # Try to connect to a Xiaomi Gateway.
+                connect_gateway_class = ConnectXiaomiGateway(self.hass)
+                await connect_gateway_class.async_connect_gateway(self.host, token)
+                gateway_info = connect_gateway_class.gateway_info
+
+                if gateway_info is not None:
+                    unique_id = (
+                        f"{gateway_info.model}-{gateway_info.mac_address}-gateway"
+                    )
+                    await self.async_set_unique_id(unique_id)
+                    self._abort_if_unique_id_configured()
+                    return self.async_create_entry(
+                        title=user_input[CONF_NAME],
+                        data={
+                            CONF_FLOW_TYPE: CONF_GATEWAY,
+                            CONF_HOST: self.host,
+                            CONF_TOKEN: token,
+                            "gateway_id": unique_id,
+                            "model": gateway_info.model,
+                            "mac": gateway_info.mac_address,
+                        },
+                    )
+
+                errors["base"] = "connect_error"
+            else:
+                errors["base"] = "no_host"
```

Comment:
This can never happen because it's part of schema?

Reply:
If zeroconf is used, the host is already set and then we use a different schema without the host (because we already now that) so then it is not in the schema.
However in principle the scheme without the host is only shown if the host is already set.
I made this error just in case to be sure, but I don't think it can ever happen.

Schould I remove it?

Reply:
Yeah

Reply:
done